test/unit/org/apache/cassandra/auth/AuthCacheInvalidationTest.java

```diff
@@ -63,146 +83,161 @@ public void prepare() throws Throwable
         executeNet("DROP KEYSPACE IF EXISTS " + ks);
 
         executeNet("CREATE USER " + user1.getRoleName() + " WITH PASSWORD '" + userPw + '\'');
-        executeNet("CREATE ROLE " + role1.getRoleName());
-        executeNet("CREATE ROLE " + role2.getRoleName());
+        createRole(role1);
+        createRole(role2);
         executeNet("GRANT " + role1.getRoleName() + " TO " + user1.getRoleName());
         executeNet("GRANT " + role2.getRoleName() + " TO " + role1.getRoleName());
 
         execute("CREATE KEYSPACE " + ks + " WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}");
         execute("CREATE TABLE " + ks + '.' + tab + " (id int PRIMARY KEY)");
         execute("CREATE TABLE " + ks + '.' + tab2 + " (id int PRIMARY KEY)");
 
-        executeNet("GRANT SELECT, UPDATE ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
-        executeNet("GRANT UPDATE ON KEYSPACE " + ks + " TO " + role2.getRoleName());
+        executeNet("GRANT SELECT ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
+        executeNet("GRANT MODIFY ON TABLE " + ks + '.' + tab + " TO " + role1.getRoleName());
+        executeNet("GRANT MODIFY ON KEYSPACE " + ks + " TO " + role2.getRoleName());
 
-        DatabaseDescriptor.getAuthManager().invalidateCaches();
+        Roles.clearCache();
+        AuthenticatedUser.clearCache();
+    }
+
+    private void createRole(RoleResource role) throws Throwable
+    {
+        executeNet(format("CREATE ROLE %s WITH LOGIN = true AND password = '%s'", role.getRoleName(), userPw));
     }
 
     @Test
     public void testInvalidationOnGrant() throws Throwable
     {
+        populateAndCheckCaches();
+        
         // Checks that the user has no SELECT permissions on tab2
-
-        useUser(user1.getRoleName(), userPw);
-
-        assertUnauthorized("SELECT * FROM " + ks + '.' + tab2,
-                           "User " + user1.getRoleName() + " has no SELECT permission on <table " + ks + '.' + tab2 + "> or any of its parents");
+        assertSelectNotAllowed(user1, ks, tab2);
+        assertSelectNotAllowed(role1, ks, tab2);
+        assertSelectNotAllowed(role2, ks, tab2);
```

Comment:
This test already had assertions about the content of the permissions cache for role1 and role2 in addition to user1.
In order to get role1/role2 entries populated into the permissions cache, they have to login and run a query.